### PR TITLE
Ndelangen/breadcrumb in header

### PR DESCRIPTION
**Pull request by ndelangen**
I'm trying to write a custom middleware for adding breadcrumbs on sourceJS pages.

I've not been able to make it work so far, I'll need some help @operatino .
It's broken at the moment.

What I think is going on: the same EJS-template is attempted to be rendered, once with part A of the data (head, header, footer), and a second time with breadcrumbs. But the first time EJS exits with an error: It cannot resolve 'breadcrumb'.

```
breadcrumb is not defined
    at eval (eval at <anonymous> (/sourcejs/node_modules/ejs/lib/ejs.js:495:12), <anonymous>:32:16)
    at returnedFn (/sourcejs/node_modules/ejs/lib/ejs.js:524:17)
    at exports.render (/sourcejs/node_modules/ejs/lib/ejs.js:345:37)
    at Object.ejs.render (/sourcejs/core/ejsWithHelpers.js:147:12)
    at /sourcejs/core/middlewares/wrap.js:116:49
    at /sourcejs/node_modules/graceful-fs/graceful-fs.js:76:16
    at tryToString (fs.js:414:3)
    at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:401:12)
Is there some option for EJS to ignore parts that it cannot resolve? And have it render the output of the earlier render with new data? I doubt this is intended use of EJS.
```

If you see some way of making this work, that be awesome! I'd be happy to continue working on this when I know how to continue.

### CODE REVIEW

**Comment by operatino** [comment]
Hi,

I've already answered to you in prev thread https://github.com/sourcejs/Source/issues/200#issuecomment-192886715, will see how I can help.

File: assets/css/project/breadcrumb.less [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+}
+.source_breadcrumb_i {
+    &:first-child {
+        margin-left: 0;
+    }
+    &:last-child {
+        margin-right: 0;
+    }
+}
+.source_breadcrumb_s {
```
Comment (operatino) [inline assets/css/project/breadcrumb.less]: Would be really cool, if you could define separator as pseudo in `.source_breadcrumb_i`, to no create new list item in the dom.

File: assets/css/project/breadcrumb.less [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    -webkit-overflow-scrolling: touch;
+}
+
+.source_breadcrumb_lst {
+    margin: 0;
+    padding: 0;
+    list-style: none;
+}
+
+.source_breadcrumb_lst > * {
```
Comment (operatino) [inline assets/css/project/breadcrumb.less]: please use more specific classnames - `.source_breadcrumb_i {}`

File: core/middlewares/breadcrumb.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+            try {
+                renderedBreadcrumps = ejs.render(template, {
+                    breadcrumb: data
+                }, {
+                    filename: templatePath
+                });
+            } catch (err) {
+                renderedBreadcrumps = '';
+                global.log.warn('Could not pre-render spec with EJS: ' + req.path, err);
```
Comment (operatino) [inline core/middlewares/breadcrumb.js]: Please define middleware name, would be really useful in the console.

File: core/middlewares/wrap.js [hunk 4]
```diff
@@ -10,12 +10,12 @@ var getHeaderAndFooter = require(global.pathToApp + '/core/headerFooter.js').get
 var specUtils = require(path.join(global.pathToApp,'core/lib/specUtils'));
 
 /**
- * Wrap rendered html from request with spec wrapper (header, footer, etc.)
```
Comment (operatino) [inline core/middlewares/wrap.js]: please revert this JSDoc formatting

File: core/middlewares/wrap.js [hunk 5]
```diff
@@ -117,7 +118,8 @@ exports.process = function (req, res, next) {
                     filename: templatePath
                 });
             } catch(err){
-                req.specData.renderedHtml = 'Error rendering Spec with EJS: ' + template;
+                console.log(err);
```
Comment (operatino) [inline core/middlewares/wrap.js]: I guess it was used for debugging? Please clean it, or add some prefix description like `console.log('wrap.js: error rendering spec with EJS', err);`

File: core/views/breadcrumb.ejs [hunk 6]
```diff
@@ -0,0 +1,14 @@
+<!-- Core header, just for example, use /public/assets/templates/breadcrumbs.inc.html -->
```
Comment (operatino) [inline core/views/breadcrumb.ejs]: clean this comment, it's not relevant

**Comment by operatino** [comment]
Good job, man! 

After few small fixes, that I've stated in the comments, your good to go! 

Please first merge with upstream branch, so appveyor CI passes, I've pushed a fix into `0.6.0-dev`, and before we merge, I'll need to ask you to squash commits into one, to not pollute the history.

**Comment by ndelangen** [comment]
I hope I did the squashing correctly, that was new to me.

**Comment by operatino** [comment]
Looks good! Thanks for the great job!

**Pull request merged**